README.md
# ctoolbenchmark
Datasets and functions to support the validation and benchmarking of Dutch carbon models


NEWS.md
# Changelog ctoolbenchark

## Version 0.1.0  2022-08-23
* First start of a package facilitating the benchmark of carbon tools in the Netherlands.



tests/testthat.R
library(testthat)
library(ctoolsbenchmark)

test_check("ctoolsbenchmark")


tests/testthat/test-omb_zwart.R
# test function organic matter decline according to Zwart et al. (2013)

require(testthat)

  # default inputs
  A_SOM_LOI = c(2,5)
  A_C_OF = c(NA,NA)
  A_CLAY_MI = 7.5

# test for two soils with only SOM input
test_that("estimation omb_zwart", {
  expect_equal(
    omb_zwart(A_SOM_LOI = c(2,5)),
    expected = c(0.23,0.33),
    tolerance = 0.01)
})

# test for two soils with both SOM and SOC input (and SOM is preferred above SOC)
test_that("estimation omb_zwart", {
  expect_equal(
    omb_zwart(A_SOM_LOI = c(2,4.5),
                A_C_OF = c(8,4.5)),
    expected = c(0.23,0.32),
    tolerance = 0.01)
})

# test for two soils with both SOM and SOC input (and SOC is preferred above SOM) and adapted duration and age
test_that("estimation omb_zwart", {
  expect_equal(
    omb_zwart(A_SOM_LOI = c(2,5),
                A_C_OF = c(8,14),
                A_CLAY_MI = c(25,1.5)),
    expected = c(0.26,0.33),
    tolerance = 0.01)
})


tests/testthat/test-omb_janssen.R
# test function organic matter decline according to Janssen (1986)

require(testthat)

  # default inputs
  # A_SOM_LOI = c(2,5)
  # A_C_OF = NA
  # iage = 10
  # duration = 10

# test for two soils with only SOM input
test_that("estimation omb_janssen", {
  expect_equal(
    omb_janssen(A_SOM_LOI = c(2,5)),
    expected = c(0.37,0.92),
    tolerance = 0.01)
})

# test for two soils with both SOM and SOC input (and SOC is preferred above SOM)
test_that("estimation omb_janssen", {
  expect_equal(
    omb_janssen(A_SOM_LOI = c(2,5),
                A_C_OF = c(8,14)),
    expected = c(0.29,0.51),
    tolerance = 0.01)
})

# test for two soils with both SOM and SOC input (and SOC is preferred above SOM) and adapted duration and age
test_that("estimation omb_janssen", {
  expect_equal(
    omb_janssen(A_SOM_LOI = c(2,5),
                A_C_OF = c(8,14),
                iage = 23,
                duration = 20),
    expected = c(0.09,0.16),
    tolerance = 0.01)
})


R/omb_janssen.R
#' Calculate the mean decomposition rate of Soil Organic Matter
#'
#' @param A_SOM_LOI (numeric) The percentage organic matter in the soil (\%).
#' @param A_C_OF (numeric) The carbon content of the soil (g / kg).
#' @param iage (numeric) The initial age of the soil organic matter (years). Default value is 17 years.
#' @param duration (numeric) The duration over which an an annual averaged decomposition rate is estimated (years). Default value is 10 years.
#'
#' @import data.table
#'
#' @references Janssen (1984) A simple method for calculating decomposition and accumulation of 'young' soil organic matter. Plant & Soil 76, 297-304.
#'
#' @export
omb_janssen <- function(A_SOM_LOI = NA_real_, A_C_OF = NA_real_, iage = 10, duration = 10) {

  # add visual bindings
  temp = cor_temp = NULL

  # extend A_C_OF when input is missing
  if(length(A_C_OF)==1 & is.na(A_C_OF[1])){A_C_OF <- rep(A_C_OF,length(A_SOM_LOI))}

  # Check inputs
  arg.length <- max(length(A_SOM_LOI), length(A_C_OF))
  checkmate::assert_numeric(A_SOM_LOI, lower = 0, upper = 100,len = arg.length)
  checkmate::assert_numeric(A_C_OF, lower = 0, upper = 1000, len = arg.length)
  checkmate::assert_numeric(iage, lower = 1, upper = 100)
  checkmate::assert_numeric(duration, lower = 1, upper = 100)

  # Collect data into a table
  dt <- data.table(A_SOM_LOI = A_SOM_LOI,
                   A_C_OF = A_C_OF,
                   value = NA_real_)

  # estimate carbon content (g/kg) from SOM (%) if only SOM is given
  # assuming that 50% of SOM consists of carbon (Pribyl, 2010)
  dt[is.na(A_C_OF), A_C_OF := A_SOM_LOI * 0.5 * 10]

  # set the anual temperature
  dt[, temp := 12]

  # add the temperature correction function
  dt[, cor_temp := ifelse(temp<=-1,0,ifelse(temp<=9,0.1*(temp+1),ifelse(temp<=27,2^((temp-9)/9),4)))]

  # estimate the carbon decomposition in 10 years
  dt[, value := A_C_OF * (1-exp(4.7*(((iage+cor_temp*duration)^-0.6)-(iage^-0.6))))/duration]

  # return value for C decomposition (g C / kg soil)
  value <- dt[, value]

  # return value
  return(value)

}


R/omb_eurofins.R
#' Calculate the mean decomposition rate of Soil Organic Matter
#'
#' @param B_SOILTYPE_AGR (character) The agricultural type of soil
#' @param A_SOM_LOI (numeric) The percentage organic matter in the soil (\%).
#' @param A_C_OF (numeric) The carbon content of the soil (g / kg).
#' @param A_N_RT (numeric) The organic nitrogen content of the soil in mg N / kg
#' @param A_N_PMN (numeric) The potentially mineralizable N pool (mg N / kg soil)
#' @param A_PH_CC (numeric) The pH-CaCl2 of the soil
#' @param duration (numeric) The duration over which an an annual averaged decomposition rate is estimated (years). Default value is 10 years.
#'
#' @import data.table
#'
#' @references Eurofins Agro (2013). Formules om OS-balans te berekenen voor NLV. Extension made by Hanegraaf & Bussink (2007) Schatting afbraaksnelheid organische stof. Vertaling van empirisch onderzoek naar nieuwe vuistregels - update 2007. NMI report B112.08, 15 pp.
#'
#' @export
omb_eurofins <- function(B_SOILTYPE_AGR, A_SOM_LOI = NA, A_C_OF = NA, A_N_RT = NA, A_PH_CC = NA, A_N_PMN = NA, duration = 10) {

  # add visual bindings
  cosfr = iagefr = temp = cor_temp = iage = NULL

  # Check inputs
  checkmate::assert_numeric(A_SOM_LOI, lower = 0, upper = 100)
  checkmate::assert_numeric(A_C_OF, lower = 0, upper = 1000)
  checkmate::assert_numeric(A_N_RT, lower = 0.1, upper = 30000)
  checkmate::assert_numeric(A_N_PMN, lower = 0, upper = 1000)
  checkmate::assert_numeric(A_PH_CC, lower = 2, upper = 10)
  checkmate::assert_subset(B_SOILTYPE_AGR,
                           choices = c("duinzand","dekzand","zeeklei","rivierklei","maasklei","dalgrond","moerige_klei", "veen","loess"),
                           empty.ok = FALSE)
  checkmate::assert_numeric(duration, lower = 1, upper = 100)

  # Collect data into a table
  dt <- data.table(B_SOILTYPE_AGR = B_SOILTYPE_AGR,
                   A_SOM_LOI = A_SOM_LOI,
                   A_C_OF = A_C_OF,
                   A_N_RT = A_N_RT * 0.001,
                   A_N_PMN = A_N_PMN,
                   A_PH_CC = A_PH_CC,
                   value = NA_real_)

  # set C-OS fraction dependent on soil type
  dt[B_SOILTYPE_AGR =='loess',cosfr := 0.44]
  dt[B_SOILTYPE_AGR %in% c("moerige_klei", "veen"),cosfr := 0.50]
  dt[B_SOILTYPE_AGR %in% c("rivierklei", "maasklei"),cosfr := 0.41]
  dt[B_SOILTYPE_AGR =='zeeklei',cosfr := 0.46]
  dt[is.na(cosfr), cosfr := 0.58]

  # add soil dependent correctionfactor for the initial age
  dt[!is.na(A_N_RT) & !is.na(A_N_PMN) & !is.na(A_PH_CC), iagefr := 3 * (-0.2366 * A_C_OF + 0.8999 * A_C_OF^0.8 - 0.01496 * A_N_PMN - 0.1722 * A_N_RT + 0.0936 * A_PH_CC)/17]
  dt[is.na(iagefr),iagefr := 1]

  # set initial age
  dt[,iage := fifelse(B_SOILTYPE_AGR == 'duinzand',12.5,17) * iagefr]

  # set the anual temperature
  dt[, temp := 14]

  # add the temperature correction function
  dt[, cor_temp := ifelse(temp<=-1,0,ifelse(temp<=9,0.1*(temp+1),ifelse(temp<=27,2^((temp-9)/9),4)))]

  # estimate carbon content (g/kg) from SOM (%) if only SOM is given
  dt[is.na(A_C_OF), A_C_OF := A_SOM_LOI * cosfr * 10]

  # estimate the carbon decomposition in 10 years
  dt[, value := A_C_OF * (1-exp(4.7*(((iage+cor_temp*duration)^-0.6)-(iage^-0.6))))/duration]

  # return value for C concentration (g C / kg soil)
  value <- dt[, value]

  # return value
  return(value)

}


R/obm_vermeulen.R
#' Calculate the mean decomposition rate of Soil Organic Matter
#'
#' @param A_SOM_LOI (numeric) The percentage organic matter in the soil (\%).
#' @param A_C_OF (numeric) The carbon content of the soil (g / kg).
#' @param A_N_RT (numeric) The total nitrogen content of the soil (mg / kg).
#' @param duration (numeric) The duration over which an an annual averaged decomposition rate is estimated (years). Default value is 10 years.
#'
#' @import data.table
#'
#' @references Vermeulen & Hendriks (1996).Bepaling van afbraaksnelheden van organische stof in laagveen: ademhalingsmetingen aan ongestoorde veenmonsters in het laboratorium. DLO-Staring centrum, Wageningen, Rapport 288, 99 pp.
#'
#' @export
omb_vermeulen <- function(A_SOM_LOI = NA_real_, A_C_OF = NA_real_, A_N_RT = NA_real_, duration = 10) {

  # add visual bindings
  A_CN_FR = NULL

  # extend A_C_OF when input is missing
  if(length(A_C_OF)==1 & is.na(A_C_OF[1])){A_C_OF <- rep(A_C_OF,length(A_SOM_LOI))}

  # Check inputs
  arg.length <- max(length(A_SOM_LOI), length(A_C_OF), length(A_N_RT))
  checkmate::assert_numeric(A_SOM_LOI, lower = 0, upper = 100,len = arg.length)
  checkmate::assert_numeric(A_C_OF, lower = 0, upper = 1000, len = arg.length)
  checkmate::assert_numeric(A_N_RT, lower = 0.1, upper = 30000, any.missing = FALSE, len = arg.length)
  checkmate::assert_numeric(duration, lower = 1, upper = 100)

  # Collect data into a table
  dt <- data.table(A_SOM_LOI = A_SOM_LOI,
                   A_C_OF = A_C_OF,
                   A_N_RT = A_N_RT * 0.001,
                   value = NA_real_)

  # estimate carbon content (g/kg) from SOM (%) if only SOM is given
  # assuming that 50% of SOM consists of carbon (Pribyl, 2010)
  dt[is.na(A_C_OF), A_C_OF := A_SOM_LOI * 0.5 * 10]

  # estimate the CN ratio, and set the applicability range (see Vermeulen & Hendriks)
  dt[, A_CN_FR := A_C_OF / A_N_RT]
  dt[, A_CN_FR := pmin(pmax(A_CN_FR,10),55)]

  # estimate the averaged annual carbon decomposition (g C kg-1 year-1), based on duration of (default) 10 years
  dt[, value := A_C_OF * (1 - exp(-1 * (0.016 - 0.00021 * A_C_OF / A_N_RT) * duration)) / duration]

  # return value for C decomposition (g C / kg soil)
  value <- dt[, value]

  # return value
  return(value)

}


R/est_bulkdensity.R
#' Calculate the bulk density given the pedotransferfunction in Dutch Fertilizer Recommendation
#'
#' This function calculates the bulk density of Dutch soils given texture and organic matter content.
#'
#' @param A_SOM_LOI (numeric) The percentage organic matter in the soil (\%).
#' @param A_CLAY_MI (numeric) The clay content of the soil (\%).
#'
#' @import data.table
#'
#' @references CBAV, 2022; CBGV, 2022
#'
#' @export
est_bulkdensity <- function(A_SOM_LOI, A_CLAY_MI) {

  # add visual bindings
  soiltype = dens.sand = dens.clay = cf = NULL

  # Check input
  arg.length <- max(length(A_SOM_LOI), length(A_CLAY_MI))
  checkmate::assert_numeric(A_SOM_LOI, lower = 0, upper = 100, any.missing = FALSE, len = arg.length)
  checkmate::assert_numeric(A_CLAY_MI, lower = 0.1, upper = 75,  len = arg.length)

  # Collect data into a table
  dt <- data.table(A_SOM_LOI = A_SOM_LOI,
                   A_CLAY_MI = A_CLAY_MI,
                   value = NA_real_
  )

  # estimate soil density (CBAV, 2019)
  # https://www.handboekbodemenbemesting.nl/nl/handboekbodemenbemesting/Bodem/Volumegewicht-grond.htm
  # https://edepot.wur.nl/413891

  # calculate soil texture dependent density (kg / m3)
  dt[, dens.sand := (1 / (0.02525 * A_SOM_LOI + 0.6541)) * 1000]
  dt[, dens.clay :=  (0.00000067*A_SOM_LOI^4 - 0.00007792*A_SOM_LOI^3 + 0.00314712*A_SOM_LOI^2 - 0.06039523*A_SOM_LOI + 1.33932206) * 1000]

  # fraction clay correction
  dt[, cf := pmin(1, A_CLAY_MI/25)]

  # clay dependent density
  dt[, value := cf * dens.clay + (1-cf) * dens.sand]

  # return value
  value <- dt[, value]

  # return value
  return(value)

}


R/omb_zwart.R
#' Calculate the mean decomposition rate of Soil Organic Matter
#'
#' @param A_SOM_LOI (numeric) The percentage organic matter in the soil (\%).
#' @param A_C_OF (numeric) The carbon content of the soil (g / kg).
#' @param A_CLAY_MI (numeric) The clay content of the soil (\%).
#'
#' @import data.table
#'
#' @references Zwart et al. (2013) De organische stof balans met de te verwachten stikstoflevering per teeltrotatie. Available at: https://edepot.wur.nl/272649.
#'
#' @export
omb_zwart <- function(A_SOM_LOI = NA_real_, A_C_OF = NA_real_,A_CLAY_MI = 5) {

  # add visual bindings
  A_DENSITY = NULL

  # extend A_C_OF when input is missing
  if(length(A_C_OF)==1 & is.na(A_C_OF[1])){A_C_OF <- rep(A_C_OF,length(A_SOM_LOI))}

  # Check inputs
  arg.length <- max(length(A_SOM_LOI), length(A_C_OF))
  checkmate::assert_numeric(A_SOM_LOI, lower = 0, upper = 100,len = arg.length)
  checkmate::assert_numeric(A_C_OF, lower = 0, upper = 1000, len = arg.length)
  checkmate::assert_numeric(A_CLAY_MI, lower = 0, upper = 100)

  # Collect data into a table
  dt <- data.table(A_SOM_LOI = A_SOM_LOI,
                   A_C_OF = A_C_OF,
                   A_CLAY_MI = A_CLAY_MI,
                   value = NA_real_)

  # estimate SOM (%) from carbon content (g/kg) if only SOM is given
  # assuming that 50% of SOM consists of carbon (Pribyl, 2010)
  dt[is.na(A_SOM_LOI), A_SOM_LOI := A_C_OF * 2 * 0.1]

  # estimate density of the soil
  dt[,A_DENSITY := est_bulkdensity(A_SOM_LOI = A_SOM_LOI, A_CLAY_MI = A_CLAY_MI)]

  # soil organic matter decomposition (kg EOS per year)
  dt[A_SOM_LOI > 23, value := 3145.513]
  dt[A_SOM_LOI <= 23, value := A_SOM_LOI * 0.1 * 0.3 * (-0.37 * log(A_SOM_LOI) + 1.8398) * (10 ^ 6) * 0.0487 * A_SOM_LOI ^ -0.453 * 0.57]

  # convert organic matter decomposition to g/kg EOS change
  dt[, value := value * 1000 / (0.3 * 100 * 100 * A_DENSITY)]

  # estimate the carbon decomposition (g C/ kg soil)
  dt[, value := value * 0.5]

  # return value for C concentration (g C / kg soil)
  value <- dt[, value]

  # return value
  return(value)

}
